### PR TITLE
Basic e2e tests for aggregated apiserver

### LINKED ISSUE #1058: Basic e2e tests for aggregated apiserver
## Feature Request

follow up: #1048 . Now that we are going to adopt an aggregated apiserver, we have to add basic e2e tests for it.

### PR DESCRIPTION
**Pull request by aylei**
### What problem does this PR solve? 
close #1058 

### What is changed and how does it work?
A test-apiserver is introduced to simulate a multiple version example of aggregated apiserver and e2e tests for it.

The new resource `pods.example.pingcap.com` is designed for e2e tests, the field, defaulting, validation and conversions are artificially designed to simulate the real world custom resource evolution.

### Check List 

Tests 

 - E2E test

Side effects

 - Breaking backward compatibility

### Does this PR introduce a user-facing change?:
 
 ```release-note
Basic e2e tests for aggregated apiserver
 ```

The generated codes are included in a separate commit, please check https://github.com/pingcap/tidb-operator/pull/1109/commits/15f9c7621b84c067f2c5fe65ad029662211a16e3 for non-generated code changes.

@cofyc @weekface @onlymellb @tennix PTAL

### CODE REVIEW

File: pkg/apiserver/storage/store.go [hunk 1]
```diff
@@ -18,27 +18,26 @@ import (
...
-	informers "github.com/pingcap/tidb-operator/pkg/client/informers/externalversions"
+	// informers "github.com/pingcap/tidb-operator/pkg/client/informers/externalversions"
 	informerv1alpha1 "github.com/pingcap/tidb-operator/pkg/client/informers/externalversions/pingcap/v1alpha1"
 	listerv1alpha1 "github.com/pingcap/tidb-operator/pkg/client/listers/pingcap/v1alpha1"
 	"k8s.io/apimachinery/pkg/api/meta"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/conversion"
 	"k8s.io/apimachinery/pkg/runtime"
-	"k8s.io/apimachinery/pkg/util/wait"
+	// "k8s.io/apimachinery/pkg/util/wait"
```
Comment (cofyc) [inline pkg/apiserver/storage/store.go]: commented imports can be removed

File: Makefile [hunk 2]
```diff
@@ -56,10 +58,14 @@ backup-manager:
 backup-docker: backup-manager
 	docker build --tag "${DOCKER_REGISTRY}/pingcap/tidb-backup-manager:latest" images/backup-manager
 
-e2e-docker-push: e2e-docker
+e2e-setup:
+	# ginkgo doesn't work with retool for Go 1.11
+	# @CGO_ENABLED=0 go get github.com/onsi/ginkgo@v1.6.0
```
Comment (cofyc) [inline Makefile]: this is not used anymor

**Comment by cofyc** [comment]
After a quick glance at the code, this looks good to me.

File: hack/test-codegen.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+scriptdir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
+
+${scriptdir}/generate-internal-groups.sh \
+  all \
+  github.com/pingcap/tidb-operator/tests/pkg/apiserver \
+  github.com/pingcap/tidb-operator/tests/pkg/apiserver/apis \
+  github.com/pingcap/tidb-operator/tests/pkg/apiserver/apis \
+  "example.pingcap.com:v1alpha1,v1beta1" \
+  --go-header-file=${scriptdir}/boilerplate.go.txt
```
Comment (cofyc) [inline hack/test-codegen.sh]: we should verify the generated code  is up to date. I created a issue for this: https://github.com/pingcap/tidb-operator/issues/1110

Comment (aylei) [reply hack/test-codegen.sh]: Thanks!

**Pull request merged**